src/m2n/GatherScatterCommunication.cpp

```diff
@@ -65,101 +66,116 @@ void GatherScatterCommunication::send(precice::span<double const> itemsToSend, i
+    utils::IntraComm::getCommunication()->receive(span<double>{secondaryRankValues}, secondaryRank);
+    for (size_t i = 0; i < secondaryDistribution.size(); ++i) {
+      utils::add_n(&secondaryRankValues[i * valueDimension], valueDimension, &globalItemsToSend[secondaryDistribution[i] * valueDimension]);
+    }
   }
+
+  // Send data to other primary
+  PRECICE_DEBUG("Sending gathered data to other participant");
+  _com->sendRange(globalItemsToSend, 0);
 }
 
 void GatherScatterCommunication::receive(precice::span<double> itemsToReceive, int valueDimension)
 {
   PRECICE_TRACE(itemsToReceive.size());
 
-  std::vector<double> globalItemsToReceive;
-
-  // Receive data at primary
-  if (not utils::IntraComm::isSecondary()) {
-    int globalSize = _mesh->getGlobalNumberOfVertices() * valueDimension;
-    PRECICE_DEBUG("Global Size = {}", globalSize);
-
-    globalItemsToReceive = _com->receiveRange(0, com::AsVectorTag<double>{});
-    PRECICE_ASSERT(globalItemsToReceive.size() == static_cast<std::size_t>(globalSize));
-  }
-
-  // Scatter data
+  // Secondary ranks receive scattered data
   if (utils::IntraComm::isSecondary()) { // Secondary rank
     if (!itemsToReceive.empty()) {
-      PRECICE_DEBUG("itemsToRec[0] = {}", itemsToReceive[0]);
       auto received = utils::IntraComm::getCommunication()->receiveRange(0, com::AsVectorTag<double>{});
+      PRECICE_ASSERT(!received.empty());
+      PRECICE_DEBUG("Received scattered data starting with {}", received[0]);
       std::copy(received.begin(), received.end(), itemsToReceive.begin());
-      PRECICE_DEBUG("itemsToRec[0] = {}", itemsToReceive[0]);
     }
-  } else { // Primary rank or coupling mode
-    PRECICE_ASSERT(utils::IntraComm::getRank() == 0);
-    mesh::Mesh::VertexDistribution &vertexDistribution = _mesh->getVertexDistribution();
-
-    // Primary rank data
-    for (size_t i = 0; i < vertexDistribution[0].size(); i++) {
-      for (int j = 0; j < valueDimension; j++) {
-        itemsToReceive[i * valueDimension + j] = globalItemsToReceive[vertexDistribution[0][i] * valueDimension + j];
-      }
+    return;
+  }
+
+  // Primary rank receives and scatters the data
+  PRECICE_ASSERT(not utils::IntraComm::isSecondary());
+
+  int globalSize = _mesh->getGlobalNumberOfVertices() * valueDimension;
+  PRECICE_DEBUG("Receiving {} elements from other participant to scatter", globalSize);
+
+  auto globalItemsToReceive = _com->receiveRange(0, com::AsVectorTag<double>{});
+  PRECICE_ASSERT(globalItemsToReceive.size() == static_cast<std::size_t>(globalSize));
+
+  const auto &vertexDistribution = _mesh->getVertexDistribution();
+
+  // Directly copy primary rank data
+  PRECICE_ASSERT(vertexDistribution.count(0) > 0);
+  const auto &primaryDistribution = vertexDistribution.at(0);
+  for (size_t i = 0; i < primaryDistribution.size(); ++i) {
+    std::copy_n(&globalItemsToReceive[primaryDistribution[i] * valueDimension], valueDimension, &itemsToReceive[i * valueDimension]);
+  }
```

Comment:
This part looks very similar to the sending part. Could we maybe add a subroutine to avoid duplication.